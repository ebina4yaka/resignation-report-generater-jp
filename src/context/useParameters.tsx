import { createContext, useCallback, useState } from 'react'
import dayjs from 'dayjs'

type ParametersContext = {
  name: string
  companyName: string
  dateOfRetirement: string
  reason: string
  daysOfPaidLeaveRemaining: number
  setName: (name: string) => void
  setCompanyName: (companyName: string) => void
  setDateOfRetirement: (dateOfRetirement: string) => void
  setReason: (reason: string) => void
  setDaysOfPaidLeaveRemaining: (daysOfPaidLeaveRemaining: number) => void
}

const defaultContext: ParametersContext = {
  name: '',
  companyName: '',
  dateOfRetirement: '',
  reason: '',
  daysOfPaidLeaveRemaining: 0,
  setName: () => {},
  setCompanyName: () => {},
  setDateOfRetirement: () => {},
  setReason: () => {},
  setDaysOfPaidLeaveRemaining: () => {},
}

export const parametersContext = createContext<ParametersContext>(
  defaultContext
)

export default function useParameters(): ParametersContext {
  const [name, setNameState] = useState('')
  const [companyName, setCompanyNameState] = useState('')
  const [dateOfRetirement, setDateOfRetirementState] = useState(
    dayjs().format('YYYY-MM-DD')
  )
  const [reason, setReasonState] = useState('')
  const [daysOfPaidLeaveRemaining, setDaysOfPaidLeaveRemainingState] = useState(
    0
  )
  const setName = useCallback((current: string): void => {
    setNameState(current)
  }, [])
  const setCompanyName = useCallback((current: string): void => {
    setCompanyNameState(current)
  }, [])
  const setDateOfRetirement = useCallback((current: string): void => {
    setDateOfRetirementState(current)
  }, [])
  const setReason = useCallback((current: string): void => {
    setReasonState(current)
  }, [])
  const setDaysOfPaidLeaveRemaining = useCallback((current: number): void => {
    setDaysOfPaidLeaveRemainingState(current)
  }, [])
  return {
    name,
    companyName,
    dateOfRetirement,
    reason,
    daysOfPaidLeaveRemaining,
    setName,
    setCompanyName,
    setDateOfRetirement,
    setReason,
    setDaysOfPaidLeaveRemaining,
  }
}
