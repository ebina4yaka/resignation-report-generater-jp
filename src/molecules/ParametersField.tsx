import React, { ChangeEvent, ReactElement, useContext, useEffect } from 'react'
import TextField from '@material-ui/core/TextField'
import InputAdornment from '@material-ui/core/InputAdornment'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { parametersContext } from '../context/useParameters'
import removeNonNumberFromString from '../libs/removeNonNumberFromString'
import calcDateOfRetirement from '../libs/calcDateOfRetirement'
import convertNumberFromZenkaku from '../libs/convertNumberFromZenkaku'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      flexWrap: 'wrap',
      flexDirection: 'column',
      justifyContent: 'center',
    },
    textField: {
      width: '30ch',
      margin: theme.spacing(1),
      marginLeft: 'auto',
      marginRight: 'auto',
    },
  })
)

export default function ParametersField(): ReactElement {
  const classes = useStyles()
  const context = useContext(parametersContext)
  const {
    name,
    companyName,
    department,
    representativeDirector,
    dateOfNotification,
    reason,
    daysOfPaidLeaveRemaining,
    dateOfRetirement,
  } = context

  useEffect(() => {
    context.setDateOfRetirement(calcDateOfRetirement(dateOfNotification))
  }, [dateOfNotification])

  const handleChangeName = (event: ChangeEvent<HTMLInputElement>) => {
    const { target } = event
    context.setName(target.value)
  }

  const handleChangeRepresentativeDirector = (
    event: ChangeEvent<HTMLInputElement>
  ) => {
    const { target } = event
    context.setRepresentativeDirector(target.value)
  }

  const handleChangeCompanyName = (event: ChangeEvent<HTMLInputElement>) => {
    const { target } = event
    context.setCompanyName(target.value)
  }

  const handleChangeDepartment = (event: ChangeEvent<HTMLInputElement>) => {
    const { target } = event
    context.setDepartment(target.value)
  }

  const handleChangeDateOfNotification = (
    event: ChangeEvent<HTMLInputElement>
  ) => {
    const { target } = event
    context.setDateOfNotification(target.value)
  }

  const handleChangeReason = (event: ChangeEvent<HTMLInputElement>) => {
    const { target } = event
    context.setReason(target.value)
  }

  const handleChangeDaysOfPaidLeaveRemaining = (
    event: ChangeEvent<HTMLInputElement>
  ) => {
    const { target } = event
    const valueNumber = Number(
      removeNonNumberFromString(convertNumberFromZenkaku(target.value))
    )
    context.setDaysOfPaidLeaveRemaining(valueNumber)
    target.value = String(valueNumber)
  }

  const handleChangeDateOfRetirement = (
    event: ChangeEvent<HTMLInputElement>
  ) => {
    const { target } = event
    context.setDateOfRetirement(target.value)
  }

  return (
    <div className={classes.root}>
      <TextField
        required
        id="name"
        label="退職者"
        defaultValue={name}
        placeholder="無職太郎"
        InputLabelProps={{
          shrink: true,
        }}
        className={classes.textField}
        autoComplete="off"
        onChange={handleChangeName}
      />
      <TextField
        required
        id="companyName"
        label="会社名"
        defaultValue={companyName}
        placeholder="株式会社無職"
        InputLabelProps={{
          shrink: true,
        }}
        className={classes.textField}
        autoComplete="off"
        onChange={handleChangeCompanyName}
      />
      <TextField
        id="department"
        label="所属部署"
        defaultValue={department}
        placeholder="開発部退職課"
        InputLabelProps={{
          shrink: true,
        }}
        className={classes.textField}
        autoComplete="off"
        onChange={handleChangeDepartment}
      />
      <TextField
        required
        id="representativeDirector"
        label="代表取締役"
        defaultValue={representativeDirector}
        placeholder="無職社長"
        InputLabelProps={{
          shrink: true,
        }}
        className={classes.textField}
        autoComplete="off"
        onChange={handleChangeRepresentativeDirector}
      />
      <TextField
        required
        id="reason"
        label="退職理由"
        defaultValue={reason}
        placeholder="一身上の都合"
        InputLabelProps={{
          shrink: true,
        }}
        className={classes.textField}
        autoComplete="off"
        onChange={handleChangeReason}
      />
      <TextField
        required
        id="dateOfNotification"
        label="退職届け出日"
        type="date"
        defaultValue={dateOfNotification}
        className={classes.textField}
        InputLabelProps={{
          shrink: true,
        }}
        autoComplete="off"
        onChange={handleChangeDateOfNotification}
      />
      <TextField
        required
        label="消化有給休暇日数"
        id="daysOfPaidLeaveRemaining"
        defaultValue={daysOfPaidLeaveRemaining}
        className={classes.textField}
        InputProps={{
          endAdornment: <InputAdornment position="end">日</InputAdornment>,
        }}
        InputLabelProps={{
          shrink: true,
        }}
        autoComplete="off"
        onChange={handleChangeDaysOfPaidLeaveRemaining}
      />
      <TextField
        required
        id="dateOfRetirement"
        label="退職日"
        type="date"
        inputProps={{ min: calcDateOfRetirement(dateOfNotification) }}
        value={dateOfRetirement}
        className={classes.textField}
        InputLabelProps={{
          shrink: true,
        }}
        autoComplete="off"
        onChange={handleChangeDateOfRetirement}
      />
    </div>
  )
}
