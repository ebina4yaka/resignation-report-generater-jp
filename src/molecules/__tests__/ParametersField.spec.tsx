import React from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render, queryByAttribute } from '@testing-library/react'
import { renderHook, act } from '@testing-library/react-hooks'
import ParametersField from '../ParametersField'
import useParameters, { parametersContext } from '../../context/useParameters'

describe('ParametersField Tests', () => {
  test('Snapshot test', () => {
    const tree = renderer.create(<ParametersField />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Input onChange Handle Test', () => {
    const { result } = renderHook(() => useParameters())
    const dom = render(
      <parametersContext.Provider value={result.current}>
        <ParametersField />
      </parametersContext.Provider>
    )
    const getById = queryByAttribute.bind(null, 'id')

    const inputName = getById(dom.container, 'name') as HTMLInputElement

    const inputCompanyName = getById(
      dom.container,
      'companyName'
    ) as HTMLInputElement
    const inputDepartment = getById(
      dom.container,
      'department'
    ) as HTMLInputElement
    const inputRepresentativeDirector = getById(
      dom.container,
      'representativeDirector'
    ) as HTMLInputElement
    const inputDateOfNotification = getById(
      dom.container,
      'dateOfNotification'
    ) as HTMLInputElement
    const inputReason = getById(dom.container, 'reason') as HTMLInputElement
    const inputDaysOfPaidLeaveRemaining = getById(
      dom.container,
      'daysOfPaidLeaveRemaining'
    ) as HTMLInputElement
    const inputDateOfRetirement = getById(
      dom.container,
      'dateOfRetirement'
    ) as HTMLInputElement

    act(() => {
      fireEvent.change(inputName, { target: { value: '無職太郎' } })
      fireEvent.change(inputCompanyName, { target: { value: '株式会社無職' } })
      fireEvent.change(inputDepartment, { target: { value: '開発部退職課' } })
      fireEvent.change(inputRepresentativeDirector, {
        target: { value: '無職社長' },
      })
      fireEvent.change(inputDateOfNotification, {
        target: { value: '2020-01-01' },
      })
      fireEvent.change(inputReason, {
        target: { value: '労働基準法に違反しているため' },
      })
      fireEvent.change(inputDaysOfPaidLeaveRemaining, {
        target: { value: '10' },
      })
      fireEvent.change(inputDateOfRetirement, {
        target: { value: '2020-01-18' },
      })
    })

    expect(result.current.name).toBe('無職太郎')
    expect(result.current.companyName).toBe('株式会社無職')
    expect(result.current.department).toBe('開発部退職課')
    expect(result.current.representativeDirector).toBe('無職社長')
    expect(result.current.dateOfNotification).toBe('2020-01-01')
    expect(result.current.reason).toBe('労働基準法に違反しているため')
    expect(result.current.daysOfPaidLeaveRemaining).toBe(10)
    expect(result.current.dateOfRetirement).toBe('2020-01-18')
  })
})
