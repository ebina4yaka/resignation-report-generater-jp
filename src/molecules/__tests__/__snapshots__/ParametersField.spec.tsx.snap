// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParametersField Tests Snapshot test 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-textField-3"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
      data-shrink={true}
      htmlFor="name"
      id="name-label"
    >
      退職者
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        defaultValue=""
        disabled={false}
        id="name"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="無職太郎"
        required={true}
        type="text"
      />
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-textField-3"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
      data-shrink={true}
      htmlFor="companyName"
      id="companyName-label"
    >
      会社名
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        defaultValue=""
        disabled={false}
        id="companyName"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="株式会社無職"
        required={true}
        type="text"
      />
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-textField-3"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink={true}
      htmlFor="department"
      id="department-label"
    >
      所属部署
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        defaultValue=""
        disabled={false}
        id="department"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="開発部退職課"
        required={false}
        type="text"
      />
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-textField-3"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
      data-shrink={true}
      htmlFor="representativeDirector"
      id="representativeDirector-label"
    >
      代表取締役
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        defaultValue=""
        disabled={false}
        id="representativeDirector"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="無職社長"
        required={true}
        type="text"
      />
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-textField-3"
    onChangeCapture={[Function]}
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
      data-shrink={true}
      htmlFor="dateOfRetirement"
      id="dateOfRetirement-label"
    >
      退職日
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        defaultValue=""
        disabled={false}
        id="dateOfRetirement"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="date"
      />
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-textField-3"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
      data-shrink={true}
      htmlFor="reason"
      id="reason-label"
    >
      退職理由
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        defaultValue=""
        disabled={false}
        id="reason"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="一身上の都合"
        required={true}
        type="text"
      />
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-textField-3"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
      data-shrink={true}
      htmlFor="daysOfPaidLeaveRemaining"
      id="daysOfPaidLeaveRemaining-label"
    >
      残有給休暇日数
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
        defaultValue={0}
        disabled={false}
        id="daysOfPaidLeaveRemaining"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="text"
      />
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          日
        </p>
      </div>
    </div>
  </div>
</div>
`;
